tooling/minitest/src/tests/slice.rs

```diff
@@ -0,0 +1,201 @@
+        f.print(load(slice_place));
+        f.exit();
+        p.finish_function(f)
+    };
+
+    let p = p.finish_program(f);
+    assert_ill_formed::<BasicMem>(p, "ValueExpr::Load: unsized value type");
+}
+
+#[test]
+fn transmute_not_wf() {
+    let mut p = ProgramBuilder::new();
+
+    let f = {
+        let mut f = p.declare_function();
+        let arr = f.declare_local::<[u32; 1]>();
+        f.storage_live(arr);
+        f.assign(index(arr, const_int(0)), const_int(42_u32));
+        // ill formed transmute:
+        let slice = transmute(load(arr), <[u32]>::get_type());
+        // (also ill formed print, but need some way to use the valueexpr)
+        f.print(slice);
+        f.exit();
+        p.finish_function(f)
+    };
+
+    let p = p.finish_program(f);
+    assert_ill_formed::<BasicMem>(p, "Cast::Transmute: unsized target type");
+}
+
+/// Tests that a wide pointer can be transmuted from a `(*T, usize)`.
+#[test]
+fn index_with_transmuted() {
+    let mut p = ProgramBuilder::new();
+
+    let f = {
+        let mut f = p.declare_function();
+        // Make array
+        let arr = f.declare_local::<[u32; 3]>();
+        f.storage_live(arr);
+        f.assign(index(arr, const_int(0)), const_int(42_u32));
+        f.assign(index(arr, const_int(1)), const_int(43_u32));
+        f.assign(index(arr, const_int(2)), const_int(44_u32));
+        let slice_ptr = ref_as_slice::<u32>(&mut f, arr, 3);
+        // Print slice[1]
+        let loaded_val = load(index(deref(load(slice_ptr), <[u32]>::get_type()), const_int(1)));
+        f.print(loaded_val);
+        f.exit();
+        p.finish_function(f)
+    };
+
+    let p = p.finish_program(f);
+    assert_eq!(get_stdout::<BasicMem>(p).unwrap(), &["43"]);
+}
+
+/// Tests that indexing into a slice throws UB for invalid indices
+#[test]
+fn invalid_index_ub() {
+    fn for_index(idx: isize) {
+        let mut p = ProgramBuilder::new();
+        let f = {
+            let mut f = p.declare_function();
+            // Make array
+            let arr = f.declare_local::<[u32; 2]>();
+            f.storage_live(arr);
+            f.assign(index(arr, const_int(0)), const_int(42_u32));
+            f.assign(index(arr, const_int(1)), const_int(43_u32));
+            let slice_ptr = ref_as_slice::<u32>(&mut f, arr, 2);
+            // This should UB
+            let loaded_val =
+                load(index(deref(load(slice_ptr), <[u32]>::get_type()), const_int(idx)));
+            f.print(loaded_val);
+            f.exit();
+            p.finish_function(f)
+        };
+        let p = p.finish_program(f);
+        assert_ub::<BasicMem>(p, "access to out-of-bounds index");
+    }
+
+    for_index(-1);
+    for_index(2);
+}
+
+fn make_slice_ptr_tuple<T: TypeConv>() -> Type {
+    tuple_ty(&[(size(0), <*mut T>::get_type()), (size(8), <u64>::get_type())], size(16), align(1))
+}
+
+fn ref_as_slice<T: TypeConv>(f: &mut FunctionBuilder, arr: PlaceExpr, known_len: u64) -> PlaceExpr {
```

Comment:
Please put helper methods at the top of the file. And at least `ref_as_slice` is non-trivial enough that it should have a doc comment.